hw/ip/otbn/util/shared/information_flow_analysis.py

```diff
@@ -0,0 +1,598 @@
+            # We shouldn't hit a LoopStart here; those cases (a loop
+            # instruction or the end of a loop) are all handled earlier
+            raise RuntimeError('Unexpected LoopStart at PC {:#x}'.format(
+                section.end))
+        elif not loc.is_special():
+            # Just a normal PC; recurse
+            result = _get_iflow(program, graph, loc.pc, constants, loop_end_pc,
+                                cache)
+
+            # Update program_end_iflow, used constants, and control_deps
+            program_end_iflow = _get_iflow_update_state(
+                result, iflow, program_end_iflow, used_constants, control_deps)
+
+            # Get information flow for return paths and new constants
+            _, rec_return_iflow, _, rec_constants, _ = result
+
+            # Update common constants
+            if common_constants is None:
+                common_constants = rec_constants
+            else:
+                for const, value in rec_constants.items():
+                    if common_constants.get(const, None) != value:
+                        # Remove from common_constants if key exists
+                        common_constants.pop(const, None)
+
+            #  Update return_iflow with the current iflow composed with return
+            #  paths
+            if rec_return_iflow is not None:
+                return_iflow = safe_update_iflow(return_iflow,
+                                                 iflow.seq(rec_return_iflow))
+        else:
+            raise RuntimeError(
+                'Unexpected next control location type at PC {:#x}: {}'.format(
+                    section.end, type(loc)))
+
+    if common_constants is None:
+        common_constants = constants
+
+    # Update the cache and return
+    out = (used_constants, return_iflow, program_end_iflow, common_constants,
+           control_deps)
+    _get_iflow_cache_update(start_pc, start_constants, out, cache)
+    return out
+
+
+def check_acyclic(graph: ControlGraph) -> None:
+    '''Checks for (non LOOP/LOOPI) cycles in control-flow graph.
+
+    If there are such cycles, we need to raise an error and not proceed; the
+    control-flow traversal would infinite-loop.
+    '''
+    cycles = graph.get_cycles(graph.start)
+    if len(cycles) != 0:
+        msg = [
+            'One or more cycles found in control-flow graph at the following PCs:'
+        ]
+        for pc, links in cycles.items():
+            msg.append('{:#x} <-> {}'.format(
+                pc, ','.join(['{:#x}'.format(l) for l in links])))
+        msg.append('Analyzing cyclic control flow outside of LOOP/LOOPI '
+                   'instructions is not currently supported.')
+        raise ValueError('\n'.join(msg))
+    return
+
+
+def get_subroutine_iflow(
+    program: OTBNProgram, graph: ControlGraph, subroutine_name: str
+) -> Tuple[Optional[InformationFlowGraph], Optional[InformationFlowGraph],
+           Dict[str, Set[int]]]:
```

Comment:
Wow! That's an impressive type annotation! :-) Possibly worth giving a name and sticking at the top?